### DISCUSSION TITLE
希望github登录能获取到用户名和头像

### DISCUSSION THREAD

**Discussion by BiFangKNT** [lobehub/lobehub#4116] [Self Hosting]
数据库版，github登录，只授权邮箱只读权限，目前只能获取邮箱地址作为用户名，且头像无法获取。理论上GitHub 用户名、头像是公开信息，不需要额外授权也能获取的。

参考说明如下：
### 使用 GitHub App 获取用户名的流程

当用户通过 GitHub App 登录时，你可以通过 OAuth 授权流程获取**安装令牌**（Installation Access Token），然后使用此令牌来访问与安装相关的信息，其中可能包括用户的 GitHub 用户名。

### 具体步骤

1. **用户授权并安装 GitHub App**：
   - 当用户在 GitHub 上安装你的 GitHub App 时，你的应用将获得与该安装关联的 `installation_id`。你需要使用该 `installation_id` 获取访问令牌。

2. **使用 `installation_id` 获取安装访问令牌**：
   - 你可以通过以下 API 调用来获取安装访问令牌，该令牌允许你访问与安装相关的 GitHub 资源（如仓库、组织）：

   ```bash
   POST https://api.github.com/app/installations/{installation_id}/access_tokens
   ```

   在这一步，你的应用会收到一个 `access_token`，这个令牌可以用于后续的 API 请求。

3. **获取用户信息**：
   - 虽然 GitHub App 的权限模型主要是针对仓库、组织等资源的操作，但你仍然可以使用用户的 `access_token` 来获取用户的公共信息，包括用户名和头像。
   - 使用以下 API 获取当前安装用户的详细信息：

   ```bash
   GET https://api.github.com/user
   ```

   这个 API 返回用户的公开信息，其中包含 GitHub 用户名、用户 ID、头像等，格式如下：

   ```json
   {
     "login": "octocat",
     "id": 1,
     "avatar_url": "https://github.com/images/error/octocat_happy.gif",
     "name": "monalisa octocat",
     "email": "octocat@github.com",
     ...
   }
   ```

### 需要注意的权限

虽然你只授权了“读取电子邮件”的权限，但 **GitHub App** 在安装过程中，可能会默认允许访问用户的公开信息，包括用户名和头像。只要你的应用安装在用户的账户上，使用安装访问令牌可以访问 `/user` 端点来获取这些信息。

### 总结

即使你只授权了电子邮件读取权限，GitHub App 的安装访问令牌仍然允许你访问用户的基本公共信息，包括用户名和头像。你可以在用户安装 GitHub App 后，通过调用 GitHub 的 `/user` API 来获取用户的用户名、ID 和头像等基本信息。这样无需增加其他权限。

**Comment by cy948** [comment]
我们的接入在OAuth scope 上与 [官方Github 接入示例](https://github.com/nextauthjs/next-auth/blob/main/packages/core/src/providers/github.ts) 一致，理论上代码是没问题的。
https://github.com/lobehub/lobe-chat/blob/4711da7793fe4d538dd2a9afde8beb4fac3e9aab/src/libs/next-auth/sso-providers/github.ts#L7-L27

**Reply by BiFangKNT** [reply]
我通过脚本确认了通过使用和您给出的实现相同的接口获取方法，我的github api是能获取到用户名和头像的。既然接口获取方法没问题，我的api也没问题，那么应该是在“把获取到的用户名和头像传递到web前端的相应绑定点中”的这个过程里出现了问题吧

测试脚本如下：
```
import os
import requests
from flask import Flask, redirect, request, url_for

app = Flask(__name__)

# 配置你的 GitHub OAuth App 的 client_id 和 client_secret
CLIENT_ID = 'your_github_client_id'
CLIENT_SECRET = 'your_github_client_secret'
# 回调 URL，必须与 GitHub OAuth App 的设置一致
REDIRECT_URI = 'http://127.0.0.1:5000/callback'

# GitHub 授权 URL
AUTH_URL = f'https://github.com/login/oauth/authorize?client_id={CLIENT_ID}&redirect_uri={REDIRECT_URI}&scope=read:user%20user:email'

# GitHub 获取访问令牌的 URL
TOKEN_URL = 'https://github.com/login/oauth/access_token'
# GitHub 用户信息 API
USER_API_URL = 'https://api.github.com/user'


# 第一步：重定向到 GitHub 授权页面
@app.route('/')
def index():
    return redirect(AUTH_URL)


# 第二步：处理 GitHub 的回调，接收授权码并用它换取访问令牌
@app.route('/callback')
def callback():
    code = request.args.get('code')  # GitHub 返回的授权码

    if not code:
        return "Error: No code provided"

    # 第三步：使用授权码请求 GitHub 的访问令牌
    token_response = requests.post(TOKEN_URL, headers={'Accept': 'application/json'}, data={
        'client_id': CLIENT_ID,
        'client_secret': CLIENT_SECRET,
        'code': code,
        'redirect_uri': REDIRECT_URI,
    })

    if token_response.status_code != 200:
        return f"Error getting token: {token_response.text}"

    # 从响应中提取访问令牌
    token_json = token_response.json()
    access_token = token_json.get('access_token')

    if not access_token:
        return "Error: No access token found"

    # 第四步：使用访问令牌请求 GitHub 用户信息
    user_response = requests.get(USER_API_URL, headers={
        'Authorization': f'token {access_token}',
        'Accept': 'application/json',
    })

    if user_response.status_code != 200:
        return f"Error fetching user info: {user_response.text}"

    # 返回用户信息
    user_info = user_response.json()
    return f"User Info: {user_info}"


if __name__ == '__main__':
    # 在本地运行应用
    app.run(debug=True)
```

**Reply by cy948** [reply]
这个不确定，你可以提一个bug issue。
